### ISSUE TITLE
Allow per-path exceptions

### ISSUE THREAD

**Issue by plynchnlm** [nodesecurity/nsp#59]
The .nsprc file currently lets me completely turn off an advisory.  Usually, I just want to turn off an advisory for one error.  For example, today I ran into advisory https://nodesecurity.io/advisories/39 for path:
      `grunt-google-cdn > google-cdn > bower > bower > handlebars > uglify-js `

In this case, the problem is that module bower needs to update module handlebars (which has already done the work of updating uglify.js to version that has a fix.)  However, adding this advisory to .nsprc would turn off the warning for all occurrences of uglify.js, of which there could be many others independent of bower.  Rather than a simple array of advisories in .nsprc, the structure could be:

    {
      "exceptions": [
         "https://nodesecurity.io/advisories/39": [
            "grunt-google-cdn > google-cdn > bower > bower > handlebars > uglify-js",
            "other > path"
          ]
         "https://nodesecurity.io/advisories/31": [ other paths...]
       ]
    }

**Comment by evilpacket** [comment]
@plynchnlm Yes that's a good point and we have discussed this current limitation before and want to make it better. I'm sure we could offer a few ways to specify exceptions, my biggest concern is making it easy for you to actually create and manage those. (although on the one hand I'm not sure exceptions should be easy as it makes it less painful to make potential issues to go away :)

@nlf any concerns with the format above?

**Comment by plynchnlm** [comment]
I just ran into a case where this would be especially helpful.  Maintainers of package bower [decided](https://github.com/bower/bower/issues/2016) to close a request to update bower's handlebars dependency to a version that did not depend on a vulnerable version of uglify.js, because they are not using the part of handlebars than runs uglify.  However, nsp will continue to report bower as containing a vulnerable uglify.js.  If I use the current exception mechanism and exclude that [vulnerability](https://nodesecurity.io/advisories/39) report, then I might miss cases of other packages which also use a vulnerable uglify.js but unlike bower really are affected by it.

**Comment by evilpacket** [comment]
@plynchnlm Thanks for the example. This is certainly something on our plate. Not sure when we will get to it but it's certainly on the list.

**Comment by wolfgang42** [comment]
I think it might be better to list all of the advisories for one path, rather than all of the paths for one advisory:
```json
{
  "path-exceptions": {
    "grunt-google-cdn > google-cdn > bower > bower > handlebars > uglify-js": [
      "https://nodesecurity.io/advisories/39",
      "https://nodesecurity.io/advisories/48"
    ],
    "other > path": [ other advisories... ]
  }
}
```

This way when a dependency is upgraded you can delete an entire block from the file, rather than having to chase up and down in the file for all of the advisories that mention that particular path.

**Comment by plynchnlm** [comment]
I don't care too much as long as there is a way to do it, but I'm not convinced that is better.  If ugliify-js is used in 5 different paths, then when it is updated with a fix to a particular vulnerability, with the approach suggested by @wolfgang42 you will still have to go to multiple places in the file to remove the exceptions.  Maybe it depends on the scenario as to which structure is better, and I'm not sure which scenario would be more common.  Perhaps nsp could support both-- though supporting just either one would be a nice improvement.

**Comment by gflarity** [comment]
I just started using NSP and I agree, this an issue for us.  I have one package that uses a vulnerable module in an optional component we don't use. The same advisory affects other modules and I do care about it. 

Also, you should be able to specify both the module and the package version. Then as the package version changes you're forced to re-evaluate your exceptions. 

This all seems straightforward enough, I'd be up for submitting a pull request but I'd like the maintainers to bless a solution/format etc so it's not a waste of effort.

**Comment by nlf** [comment]
this is something that we have plans for in the works. thanks for the offer, though! we're making lots of changes to both incoming and outgoing data when performing a check that will allow us to do things like this in a more friendly way.

**Comment by blueimp** [comment]
I had the same requirement and wrote this wrapper script which allows to filter out vulnerabilities based on path packages:

```js
#!/usr/bin/env node

'use strict'

const path = require('path')
const nsp = require('nsp')
const stdout = process.stdout

const payload = {
  shrinkwrap: path.join(__dirname, '/npm-shrinkwrap.json'),
  package: path.join(__dirname, '/package.json')
}

// Filters nsp reports based on nspIgnoreDependencies array in package.json
function filterReports (reports) {
  if (!reports || !reports.length) return []
  const ignoreDeps = require(payload.package).nspIgnoreDependencies || []
  const filteredReports = []
  reportLoop: // eslint-disable-line no-labels
    for (let report of reports) {
      for (let dependency of ignoreDeps) {
        if (report.path.includes(dependency)) {
          continue reportLoop // eslint-disable-line no-labels
        }
      }
      filteredReports.push(report)
    }
  return filteredReports
}

nsp.check(payload, function (err, data) {
  const reports = filterReports(data)
  stdout.write(nsp.formatters.default(err, reports) + '\n')
  if (reports.length || err) process.exit(1)
})
```

The script makes use of a `nspIgnoreDependencies` array in `package.json` that uses the same format as the path components for its items, e.g.:

```json
{
  "nspIgnoreDependencies": [
    "karma@1.1.2"
  ]
}
```

I hope it's of some use to someone. :relaxed:
